### ISSUE TITLE
false positive with lint "assigned but may not be used"

### ISSUE THREAD

**Issue by IndrajeetPatil** [jimhester/lintr#358]
I keep getting the following [lint](https://github.com/IndrajeetPatil/ggstatsplot/commit/d6d0e9e4b49e72cc364f71e2bc8d9b169758b379#commitcomment-31337708)
![image](https://user-images.githubusercontent.com/11330453/48663483-fa2eda00-ea5e-11e8-8b81-72fdfc08a41a.png)

But if you look at the code, this object is defined at `L361` and used immediately on `L368`.
https://github.com/IndrajeetPatil/ggstatsplot/blob/d6d0e9e4b49e72cc364f71e2bc8d9b169758b379/R/ggcorrmat.R#L361-L379

Is there any way `lintr`'s behavior can be changed to get rid of such false positives?

**Comment by stufield** [comment]
I have also seen such false positives. Especially in the RStudio IDE warnings in the margin. However, they seem less common in the preview development version of RStudio, so I'm hopeful with the coming release of RStudio 1.2 these will resolve themselves. As for `lintr` I do not know.

**Comment by russHyde** [comment]
I'm struggling to make a test that covers this example. The failing code suggests a test like:

```
expect_lint("x <- 123\nbquote(atop(1, .(x)))", NULL, object_usage_linter)
```

But that doesn't fail using the current master https://github.com/jimhester/lintr/commit/a465cd781da94142e55fded840911a98c15d5da1

Could you confirm which version of lintr you are using please @IndrajeetPatil

**Comment by russHyde** [comment]
Thanks. 
Here's a failing test based on your code:

```
expect_lint("function(){\n  x <- 123\n  bquote(atop(1, .(x)))\n}\n", NULL, object_usage_linter)

Error: got 1 lints instead of 0
list(filename = "/tmp/RtmpHMgmUs/file13e519c67579", line_number = 2, column_number = 3, type = "warning", message = "local variable ‘x’ assigned but may not be used", line = "  x <- 123", ranges = list(c(3, 3)), linter = "object_usage_linter")
```

**Comment by russHyde** [comment]
@IndrajeetPatil I'm afraid I don't know. could you try running the example failure or your original code through the current master.

**Comment by russHyde** [comment]
That's funny. I just ran the reprex, and it throws an `object_usage_linter` lint for me. Reopening.

**Comment by IndrajeetPatil** [comment]
You are right! I can see it now. Not sure why one of the Travis builds did not generate any lints.

At any rate, I am getting a new lint with the new changes to `lintr` about cyclomatic complexity and I am not sure how to get rid of this lint.  

![image](https://user-images.githubusercontent.com/11330453/64046511-01306000-cb3a-11e9-91d7-fbf190a28ee5.png)

Is there any way I can compute this index locally using `lintr`?

**Comment by russHyde** [comment]
I think you could run it from cyclocomp. Or locally update to the latest github version of lintr.

we didn't touch the `object_usage_linter` in the recent change. that looks like the next big  debug, since it has cropped up in a couple of issues.

**Comment by russHyde** [comment]
This (and #352 ) seems to relate to `codetools::checkUsage`, which is used by the `object_usage_linter`.

```
g <- function() {x <- 123; bquote(atop(1, .(x)))}
codetools::checkUsage(g)
<anonymous>: local variable ‘x’ assigned but may not be used (:1)
```

**Comment by russHyde** [comment]
@IndrajeetPatil  could you confirm whether your package passes `R CMD check`, please?

To fix this would either require 

- modifying `codetools` (which is a very established package, and for which the issue-relevant function is ran during `R CMD check`); 
- writing some functions or a wrapper package around `codetools` to distinguish things like `glue::glue('{x}')` and `bquote(atop(1, .(x)))` where a defined variable is used in an nonstandard / indirect way that codetools doesn't currently cope with
- or finding a codetools-independent solution, eg, using the parse tree for each function

The problem with the second or third option, is that they could involve loads of work and still be futile: if your package contains code of this form, (after implementing the changes to lintr) you may run lintr on that package and collect no lints, but the package would still fail R CMD check. Ultimately meaning that you'll have to change the relevant code anyway.

**Comment by jimhester** [comment]
Yes, in general both lintr and codetools are going to have trouble with non-standard evaluation, these false positives can't really be helped in the general case.

**Comment by jimhester** [comment]
The object usage linter now respects variables declared in `utils::globalVariables()`, so you can add them there to avoid these warnings.

**Comment by jimhester** [comment]
This is also generally the best way to deal with false positive `R CMD check` notes due to NSE in packages as well.
